packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1215,32 +1270,161 @@ export function errorHydratingContainer(parentContainer: Container): void {
+    validateDOMNesting(type, null, hostContextDev.ancestorInfo);
+  }
+  const ownerDocument = getOwnerDocumentFromRootContainer(
+    rootContainerInstance,
+  );
+  // For the three persistent Host Components that exist in DOM it is necessary for there to
+  // always be a documentElement. With normal html parsing this will always be the case but
+  // with pathological manipulation the document can end up in a state where no documentElement
+  // exists. We create it here if missing so we can treat it as an invariant.
+  // It is important to note that thi dom mutation and others in this function happen
+  // in render rather than commit. This is tolerable because they only happen in degenerate cases
+  if (!ownerDocument.documentElement) {
+    ownerDocument.append(ownerDocument.createElement('html'));
+  }
+  switch (type) {
+    case 'html': {
+      // We ensure this exists just above
+      return ((ownerDocument.documentElement: any): HTMLHtmlElement);
+    }
+    case 'head': {
+      if (!ownerDocument.head) {
+        ownerDocument.insertBefore(
+          ownerDocument.createElement('head'),
+          ownerDocument.firstChild,
+        );
+      }
+      // We ensure this exists just above
+      return ((ownerDocument.head: any): HTMLHeadElement);
+    }
+    case 'body': {
+      if (!ownerDocument.body) {
+        ownerDocument.appendChild(ownerDocument.createElement('body'));
+      }
+      // We ensure this exists just above
+      return ((ownerDocument.body: any): HTMLBodyElement);
+    }
+    default: {
+      throw new Error(
+        'acquireSingletonInstance was called with an element type that is not supported. This is a bug in React.',
+      );
+    }
+  }
 }
 
-export function prepareScopeUpdate(
-  scopeInstance: ReactScopeInstance,
+export function commitSingletonPlacement(
+  type: string,
+  props: Props,
+  instance: Instance,
   internalInstanceHandle: Object,
 ): void {
-  if (enableScopeAPI) {
-    precacheFiberNode(internalInstanceHandle, scopeInstance);
+  switch (type) {
+    case 'html':
+    case 'head':
+    case 'body': {
+      clearSingletonInstance(instance);
+      resetProperties(instance, type, props);
+      precacheFiberNode(internalInstanceHandle, instance);
+      updateFiberProps(instance, props);
+      return;
+    }
+    default: {
+      throw new Error(
+        'commitSingletonPlacement was called with an element type that is not supported. This is a bug in React.',
+      );
+    }
   }
 }
 
-export function getInstanceFromScope(
-  scopeInstance: ReactScopeInstance,
-): null | Object {
-  if (enableScopeAPI) {
-    return getFiberFromScopeInstance(scopeInstance);
+export function getInsertionEdge(parent: Instance): ?InstanceSibling {
+  if (enableFloat) {
+    let node = null;
+    let nextNode = parent.lastChild;
+    let fallbackNode;
+    while (nextNode != null) {
+      const fiber = getInstanceFromNodeDOMTree(nextNode);
```

Comment:
The host configs should have zero knowledge of Fibers. The cases that use them are either Scopes (which won't land in its current form) and events which will be ported away from synthetic events. So we have a reasonable trajectory to making that true.

We shouldn't add more. So we need to figure out a way to layer this better so it's minimally invasive.